source/Halibut.Tests/ExceptionContractFixture.cs

```diff
@@ -0,0 +1,228 @@
+        {
+            var halibutTimeoutsAndLimits = new HalibutTimeoutsAndLimitsForTestsBuilder().Build();
+            halibutTimeoutsAndLimits.TcpClientConnectTimeout = TimeSpan.FromSeconds(60);
+            halibutTimeoutsAndLimits.ConnectionErrorRetryTimeout = TimeSpan.FromSeconds(600);
+            halibutTimeoutsAndLimits.RetryCountLimit = int.MaxValue;
+
+            await using var clientAndService = await clientAndServiceTestCase.CreateTestCaseBuilder()
+                .AsLatestClientAndLatestServiceBuilder()
+                .NoService(false)
+                .WithStandardServices()
+                .WithHalibutTimeoutsAndLimits(halibutTimeoutsAndLimits)
+                .Build(CancellationToken);
+
+            var client = clientAndService.CreateAsyncClient<IEchoService, IAsyncClientEchoServiceWithOptions>();
+
+            var cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromSeconds(5));
+
+            (await AssertException.Throws<TaskCanceledException>(async () => await client.SayHelloAsync("Hello", new(cancellationTokenSource.Token, CancellationToken.None))))
+                .And.Message.Should().Contain("A task was canceled.");
+        }
+
+        
+// TODO: ASYNC ME UP!
+// net48 does not support cncellation of the request as the DeflateStream ends up using Begin and End methods which don't get passed the cancellation token
+#if !NETFRAMEWORK
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testNetworkConditions: false, testPolling: false, testWebSocket: false)]
+        public async Task WhenTheListeningRequestIsCancelledWhileInProgress_ATaskCanceledExceptionShouldBeThrown(ClientAndServiceTestCase clientAndServiceTestCase)
+        {
+            var executingSemaphore = new SemaphoreSlim(0, 1);
+            var waitSemaphore = new SemaphoreSlim(0, 1);
+            
+            await using var clientAndService = await clientAndServiceTestCase.CreateTestCaseBuilder()
+                .AsLatestClientAndLatestServiceBuilder()
+                .WithDoSomeActionService(() =>
+                {
+                    executingSemaphore.Release();
+                    waitSemaphore.Wait(CancellationToken);
+                })
+                .Build(CancellationToken);
+
+            var doSomeActionClient = clientAndService.CreateAsyncClient<IDoSomeActionService, IAsyncClientDoSomeActionServiceWithOptions>();
+
+            var cancellationTokenSource = new CancellationTokenSource();
+            
+            var cancellationTask = Task.Run(async () =>
+            {
+                await executingSemaphore.WaitAsync(CancellationToken);
+                cancellationTokenSource.Cancel();
+            });
+
+            // CONTRACT BUG - This will be interpreted as a Transient Error in Server and cause the incorrect behaviour on cancellation
```

Comment:
This was why Cooperative cancellation had unexpected side effects in Server 😭 

@sburmanoctopus it was a pre-existing bug from adding cancellation to the socket!!